datafusion/expr/src/accumulator.rs

```diff
@@ -44,3 +44,27 @@ pub trait Accumulator: Send + Sync + Debug {
     /// returns its value based on its current state.
     fn evaluate(&self) -> Result<ScalarValue>;
 }
+
+#[derive(Debug)]
+pub enum AggregateState {
```

Comment:
This is a very elegant idea. Can you please add docstrings to `AggregateState` explaining what is going on?

 I think it would be worth updating the docstrings in the accumulator trait with some discussion / examples of how to use the Array state.